File: dehaze-python/app/models/schema/package.py
from typing import Literal

from pydantic import BaseModel, Field


class BenefitOverrides(BaseModel):
    monthlyDehazeQuota: int | None = None
    monthlyEvaluateQuota: int | None = None
    historyRetention: int | None = None
    batchLimit: int | None = None
    priority: int | None = None
    advancedParams: int | None = None
    hdExport: int | None = None
    reportExport: int | None = None
    batchDownload: int | None = None


class PackageForm(BaseModel):
    id: int | None = None
    name: str = Field(..., min_length=2, max_length=32, description="套餐名称")
    packageType: Literal["vip", "credit"] = Field(default="vip", description="商品类型")
    creditAmount: int | None = Field(default=None, ge=1, description="可得积分(积分卡必填)")
    levelCode: str | None = None
    period: Literal["monthly", "quarterly", "yearly"] | None = None
    periodDays: int | None = Field(default=None, ge=1, le=365, description="有效期天数(积分卡为NULL)")
    originalPrice: int = Field(..., ge=1, description="原价(分)")
    salePrice: int = Field(..., ge=1, description="促销价(分)")
    description: str | None = Field(default=None, max_length=256, description="套餐描述")
    benefitOverrides: BenefitOverrides | None = None
    sort: int | None = Field(default=0, ge=0, le=999, description="排序值")
    status: int | None = Field(default=0, ge=0, le=1)


class CouponForm(BaseModel):
    id: int | None = None
    name: str = Field(..., min_length=1, description="优惠券名称")
    type: str = Field(..., description="类型")
    faceValue: int = Field(..., ge=0, description="面值")
    threshold: int | None = Field(default=None, ge=0, description="使用门槛(分)")
    validType: Literal["fixed", "relative"] = Field(..., description="有效期类型")
    validStart: str | None = None
    validEnd: str | None = None
    validDays: int | None = Field(default=None, ge=1, description="领取后有效天数")
    totalQty: int = Field(..., ge=0, description="发放总量")
    perUserLimit: int = Field(..., ge=1, description="每人限领数量")
    applicableScope: list[int | str] | None = Field(
        default=None, description="适用商品(商品ID或商品类型vip/credit，NULL全部适用)"
    )
    status: int | None = Field(default=1, ge=0, le=1)


class CouponBatchDistributeForm(BaseModel):
    couponId: int = Field(..., description="优惠券ID")
    targetScope: str = Field(..., description="发放范围(all/level/users)")
    levelCodes: list[str] | None = None
    userIds: list[int] | None = None

File: dehaze-go/internal/model/bo/package.go
package bo

// BenefitOverrides 套餐权益覆盖项。字段集须与 python `BENEFIT_FIELDS`（`service/package_service.py`）
// 及 java `model.form.BenefitOverrides` 完全一致——CRUD 保存走 json.Marshal/Unmarshal 该结构，
// 少一个字段就意味着该 key 在保存时被静默丢弃（不报错），而履约侧仍会按 key 读取它。
// 无 binding 约束：python 该模型是裸 `int | None`，不得"顺手"加范围校验。
type BenefitOverrides struct {
	MonthlyDehazeQuota          *int `json:"monthlyDehazeQuota"`
	MonthlyDerainQuota          *int `json:"monthlyDerainQuota"`
	MonthlyDesnowQuota          *int `json:"monthlyDesnowQuota"`
	MonthlyLowlightQuota        *int `json:"monthlyLowlightQuota"`
	MonthlySuperResolutionQuota *int `json:"monthlySuperResolutionQuota"`
	MonthlyDenoiseQuota         *int `json:"monthlyDenoiseQuota"`
	MonthlyInpaintQuota         *int `json:"monthlyInpaintQuota"`
	MonthlyEvaluateQuota        *int `json:"monthlyEvaluateQuota"`
	AiCreditsDaily              *int `json:"aiCreditsDaily"`
	AiCreditsMonthly            *int `json:"aiCreditsMonthly"`
	HistoryRetention            *int `json:"historyRetention"`
	BatchLimit                  *int `json:"batchLimit"`
	Priority                    *int `json:"priority"`
	AdvancedParams              *int `json:"advancedParams"`
	HdExport                    *int `json:"hdExport"`
	ReportExport                *int `json:"reportExport"`
	BatchDownload               *int `json:"batchDownload"`
}

type PackageForm struct {
	ID            int64  `json:"id"`
	Name          string `json:"name" binding:"required,min=2,max=32"`
	PackageType   string `json:"packageType"`
	CreditAmount  *int64 `json:"creditAmount"`
	LevelCode     string `json:"levelCode"`
	Period        string `json:"period"`
	PeriodDays    *int   `json:"periodDays" binding:"omitempty,min=1,max=365"`
	OriginalPrice int64  `json:"originalPrice" binding:"required,min=1"`
	SalePrice     int64  `json:"salePrice" binding:"required,min=1"`
	Description   string `json:"description" binding:"omitempty,max=256"`
	// vip/credit 差异字段的条件必填（levelCode/period/periodDays、creditAmount）由服务层 validatePackageForm 校验
	BenefitOverrides *BenefitOverrides `json:"benefitOverrides"`
	Sort             *int              `json:"sort" binding:"omitempty,min=0,max=999"`
	Status           *int              `json:"status" binding:"omitempty,oneof=0 1"`
}

type PackageStatusForm struct {
	Status int `json:"status"`
}

type CouponForm struct {
	ID           int64   `json:"id"`
	Name         string  `json:"name" binding:"required"`
	Type         string  `json:"type" binding:"required,oneof=full_reduction discount no_threshold trial"`
	FaceValue    int64   `json:"faceValue" binding:"gte=0"`
	Threshold    *int64  `json:"threshold" binding:"omitempty,gte=0"`
	ValidType    string  `json:"validType" binding:"required,oneof=fixed relative"`
	ValidStart   *string `json:"validStart"`
	ValidEnd     *string `json:"validEnd"`
	ValidDays    *int    `json:"validDays" binding:"omitempty,gte=1"`
	TotalQty     int     `json:"totalQty" binding:"gte=0"`
	PerUserLimit int     `json:"perUserLimit" binding:"gte=1"`
	// 适用商品：商品 ID 或商品类型 vip/credit（python list[int|str]），NULL 全部适用
	ApplicableScope []interface{} `json:"applicableScope"`
	Status          *int          `json:"status" binding:"omitempty,oneof=0 1"`
}

type CouponBatchDistributeForm struct {
	CouponID    int64    `json:"couponId"`
	TargetScope string   `json:"targetScope"`
	LevelCodes  []string `json:"levelCodes"`
	UserIDs     []int64  `json:"userIds"`
}
